server/odc-service/src/main/java/com/oceanbase/odc/service/db/DBPLModifyHelper.java

```diff
@@ -0,0 +1,167 @@
+import com.oceanbase.tools.dbbrowser.model.DBObjectType;
+
+import lombok.NonNull;
+
+/**
+ * @description:
+ * @author: zijia.cj
+ * @date: 2024/10/17 18:59
+ * @since: 4.3.3
+ */
+@Validated
+@Component
+@SkipAuthorize("permission check inside")
+public class DBPLModifyHelper {
+    public static final String ODC_TEMPORARY_PROCEDURE = "_ODC_TEMPORARY_PROCEDURE";
+    public static final String ODC_TEMPORARY_TRIGGER = "_ODC_TEMPORARY_TRIGGER";
+    public static final String ODC_TEMPORARY_FUNCTION = "_ODC_TEMPORARY_FUNCTION";
+    public static final String OB_VERSION_SUPPORT_MULTIPLE_SAME_TRIGGERS = "4.2";
+    public static final int LOCK_TIMEOUT_SECONDS = 3;
+    @Autowired
+    private ConnectConsoleService connectConsoleService;
+    @Autowired
+    private JdbcLockRegistry jdbcLockRegistry;
+    @Autowired
+    private ConnectSessionService sessionService;
+
+    public EditPLResp editPL(@NotNull String sessionId, @NotNull @Valid EditPLReq editPLReq)
+            throws Exception {
+        DBObjectType plType = editPLReq.getObjectType();
+        switch (plType) {
+            case PROCEDURE:
+                return executeWrappedEditPL(sessionId, editPLReq, DBObjectType.PROCEDURE, ODC_TEMPORARY_PROCEDURE);
+            case FUNCTION:
+                return executeWrappedEditPL(sessionId, editPLReq, DBObjectType.FUNCTION, ODC_TEMPORARY_FUNCTION);
+            case TRIGGER:
+                if (VersionUtils.isLessThan(
+                        ConnectionSessionUtil.getVersion(sessionService.nullSafeGet(sessionId, true)),
+                        OB_VERSION_SUPPORT_MULTIPLE_SAME_TRIGGERS)) {
+                    throw new BadRequestException(
+                            "editing trigger in mysql mode is not supported in ob version less than "
+                                    + OB_VERSION_SUPPORT_MULTIPLE_SAME_TRIGGERS);
+                }
+                return executeWrappedEditPL(sessionId, editPLReq, DBObjectType.TRIGGER, ODC_TEMPORARY_TRIGGER);
+            default:
+                throw new IllegalArgumentException(
+                        String.format("the pl type %s of editing procedure is not supported", plType));
+        }
+    }
+
+    private EditPLResp executeWrappedEditPL(String sessionId, EditPLReq editPLReq,
+            DBObjectType plType, String tempPlName) throws Exception {
+        String plName = editPLReq.getObjectName();
+        String editPLSql = editPLReq.getSql();
+        String tempPLSql = editPLSql.replaceFirst(plName, tempPlName);
+        StringBuilder wrappedSqlBuilder = new StringBuilder();
+        wrappedSqlBuilder.append("DELIMITER $$\n")
+                .append(tempPLSql).append(" $$\n")
+                .append("drop ").append(plType).append(" if exists ").append(tempPlName).append(" $$\n")
+                .append("drop ").append(plType).append(" if exists ").append(plName).append(" $$\n")
+                .append(editPLSql).append(" $$\n")
+                .append("DELIMITER ;");
```

Comment:
why should we set the delimiter to ';'?

we should record the current delimiter, and set to it when we finish the sql execution

Reply:
ok